rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,275 @@
+into individul words and phrases (lexical analysis), and collapsing 
+variations of words into a single index term (stemming.)
+
+## Goals & Motivation
+
+The current set of string filters available in our GraphQL API is lacking 
+the tools needed to build effective, modern interfaces. Wildcard string 
+matching does provide string filtering, but users have come to expect the 
+easy to use filtering that comes with fulltext search systems.
+
+To facilitate building effective user interfaces we must provide human-user 
+friendly query filtering. Lexical, composite fulltext search filters can 
+provide the tools necessary for front-end developers to implement powerful 
+search bars that filter data across multiple fields of an Entity.
+
+With the proposed feature I aim to:
+  - provide tools for subgraph developers to define composite search 
+    indexes that can search across multiple fields and entities, and
+  - integrate powerful, user friendly text search that is flexible and 
+    can be easily defined by a subgraph developer ultimately providing 
+    the tools necesssary for front end developers to create effective 
+    user interfaces without the need to middleware.
+     
+## Urgency
+
+A delay in adding the fulltext search feature  will not create issues with 
+current deployments. However, it will represent a realization of part of 
+the long term vision for the query network. Also several high profile users 
+have communicated that it may be a conversion blocker, so implementation 
+should be prioritized. 
+
+## Terminology
+
+- _lexeme_: a basic lexical unit of a language, consisting of one word or 
+  several words, considered as an abstract unit, and applied to a family 
+  of words related by form or meaning.
+- _morphology (linguistics)_: the study of words, how they are formed, 
+  and their relationship to other words in the same language. 
+- _fulltext search index_: the result of lexical and morphological 
+  analysis (stemming) of a set of text documents.  It provides frequency 
+  and location for the language-specific stems found in the text documents 
+  being indexed. 
+- _ranking algorithm_: "Ranking attempts to measure how relevant documents 
+  are to a particular query, so that when there are many matches the most 
+  relevant ones can be shown first." [- Postgres Documentation](https://www.postgresql.org/docs/11/textsearch-controls.html#TEXTSEARCH-RANKING-SEARCH-RESULTS)
+  
+  **Algorithms**:
+  - _standard ranking_: ranking based on the number of matching lexemes.     
+  - _cover density ranking_: Cover density is similar to the standard 
+    fulltext search ranking except that the proximity of matching lexemes 
+    to each other is taken into consideration. This function requires 
+    lexeme positional information to perform its calculation, so it ignores 
+    any "stripped" lexemes in the index.            
+
+## Detailed Design
+
+### Subgraph Definition
```

Comment:
Definition -> Schema. `Subgraph Definition` is not a term we use (other than informally perhaps).